### PR TITLE
Add stub of new architecture overview

### PR DESCRIPTION
**Pull request by brainwane**
@lgh2 interviewed @ewdurbin and got down some rough notes that I've polished a bit. I know this is just a sketch and not very detailed, but I suggest we review this for accuracy and get this merged, and then iteratively improve this page as Laura and other newer contributors learn more and use their experiences to add to it.

It's probably best to review this commit-by-commit so you see the `git mv` of the legacy URL structure doc separately from the new architecture overview.

Partially gets at #2794 but not entirely.

### CODE REVIEW

File: docs/application.rst [hunk 1]
```diff
@@ -1,80 +1,75 @@
...
+Warehouse's front end uses Jinja2 templates.
 
-At the moment it *just* lists the legacy structure and none of the intended
-new structure.
+The application exists within two `Docker <https://docs.docker.com/>`__
+containers, one of which contains static files for the website, and the
+other which contains the Python web application code running in a
+virtual environment and the database. In the development environment,
+`Docker Compose <https://docs.docker.com/compose/overview/>`__ manages
+running the containers and the connections between them.
```
Comment (di) [inline docs/application.rst]: Technically this isn't true _yet_ -- we currently don't use any containers in production, and we use [far more than two containers in development](https://github.com/pypa/warehouse/blob/master/docker-compose.yml#L3), but this will be true some day in the near future.

It might be worth disambiguating the difference between how we use containers in a development environment vs. how we (will eventually) use them in production.

File: docs/application.rst [hunk 2]
```diff
@@ -1,80 +1,75 @@
...
-  - `rss` and `packages_rss` will be in a bunch of peoples` RSS readers
-  - `doap` is most likely referred to
-  - `show_md5` ?
+Directories within the repository:
 
-**can be deprecated carefully**
-  - `submit`, `upload`, `doc_upload`, `file_upload`,
+::
+
+    bin/ - high-level scripts for Docker
```
Comment (di) [inline docs/application.rst]: Also used for Travis and Makefile commands

File: docs/application.rst [hunk 3]
```diff
@@ -1,80 +1,75 @@
...
+        legacy/ - most of the read-only APIs implemented here
+        forklift/ - APIs for upload
+        accounts/ - user accounts
+        admin/ - application-administrator-specific
+        cache/ - caching
+        classifiers/ - frame trove classifiers
+        cli/ - entry scripts and [the interactive shell](https://warehouse.readthedocs.io/development/getting-started/#running-the-interactive-shell)
+        i18n/ - internationalization
+        locales/ - internationalization
+        manage/ - DB
```
Comment (di) [inline docs/application.rst]: This is actually for everything a logged-in user can do, (i.e. manage account and owned projects)

File: docs/application.rst [hunk 4]
```diff
@@ -1,80 +1,75 @@
...
+        migrations/ - DB
+        packaging/ - models
+                    - rate limiting to prevent abuse
+                    - RSS feeds
+                    - site maps
```
Comment (di) [inline docs/application.rst]: Seems like the directory names are missing here.

File: docs/application.rst [hunk 5]
```diff
@@ -1,80 +1,75 @@
...
-/raw-packages nginx implementation specific hackery (entirely internal; not to
-              be retained)
-/stats        Web stats. Whatever. Probably dead.
-/local-stats  Package download stats. All the legacy mirrors have this.
-/static       Static files (CSS, images) in support of the web interface.
-============= =================================================================
+Warehouse also uses `Pyramid's hybrid URL traversal and
+dispatch <https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/hybrid.html>`__.
+Using factory classes, URLs are pre-populated before the view is
+requested.
```
Comment (di) [inline docs/application.rst]: I would rephrase this line a bit, I'm not sure 'URLs are pre-populated' makes sense, maybe something like "Using factory classes, resources are provided directly to the views based on the URL pattern"

File: docs/application.rst [hunk 6]
```diff
@@ -1,80 +1,75 @@
...
+-  ``LICENSE``
+-  ``CONTRIBUTING.rst`` (the contribution guide)
+-  ``README.rst``
+-  ``requirements.txt`` for the Warehouse virtual environment
+-  ``Dockerfile``: creates the Docker containers that Warehouse runs in
+-  ``docker-compose.yml`` file configures Docker Compose
+-  ``setup.cfg`` for test configuration
+-  ``runtime.txt`` for Heroku
+-  ``Makefile``: commands to spin up Docker Compose and the Docker
+   containers
```
Comment (di) [inline docs/application.rst]: And run tests, and linting, etc.

**Comment by di** [comment]
Also @brainwane you've got a bad doc reference that's failing CI:
```
/home/travis/build/pypa/warehouse/docs/development/legacy-application-structure.rst:20:unknown document: api-reference/legacy
```

**Comment by brainwane** [comment]
Thanks @di. I've fixed all those issues.
